File: Reassortment/comp1.3.py
# Influenza Competition model 1.2.2
# Same concept as comp model 1 but more efficient. 
# Some key differences:
# - back mutation, tracking info, and program cut strategy parameterized.
# - No sequence information using array. Only keeps track of mutant allele amount

import numpy as np
import datetime
import timeit
from progress.bar import Bar
import sys
import os

# Parameters
# N0 = Population size
# K = Carrying capacity
# L = sequence length
# s = fitness decrease from deleterious mutation
# mu = mutation rate per site
# gen_num = generation amount 
# r = reassortment rate
# rep = repetition amount
# N1r = ratio of 1segment virus

#default parameters
back = 0
timestep = 0
krecord = 0
rep = 1
L = 300
s = 0.05
N0 = 1000
K = 1000
mu = 0.0005
gen_num = 10
cost = 0.00
r = 0.5
N1r = 0.5

if len(sys.argv) > 1: # input parameters from command line
    try:
        back = int(sys.argv[2])
        timestep = int(sys.argv[3])
        krecord = int(sys.argv[4])
        rep = int(sys.argv[5])
        L = int(sys.argv[6])
        s = float(sys.argv[7])
        N0 = int(sys.argv[8])
        K = int(sys.argv[9])
        mu = float(sys.argv[10])
        gen_num = int(sys.argv[11])
        cost = float(sys.argv[12])
        r = float(sys.argv[13])
        N1r = float(sys.argv[14])
    except:
        pass

class Virus1():
    """
    This class produces objects which are single agents of a influenza virus with 1 segment.
    k = number of deleterious mutation in entire genome
    s = fitness decrease from deleterious mutation
    L = sequence length for a virus
    cost = cost of having multisegments
    w = fitness
    """
    def __init__(self,k):
    	self.id = 1
        self.k = k
    
    def mutate(self,mu):
        """
        Mutation in sequence before reproduction
        mu = mutation rate
        """
        mut_num = int(np.random.binomial(L, mu)) # number of mutation
        if not back:
            self.k += mut_num
        else:
            for i in range(mut_num):
                if np.random.uniform(0,L) < self.k:
                    self.k -= 1 # back mutation
                else:
                    self.k += 1 # normal mutation
                    
class Virus2():
    """
    This class produces objects which are single agents of a influenza virus with 2 segments.
    k1 = number of deleterious mutation in segment1
    k2 = number of deleterious mutation in segment2
    k = number of deleterious mutation in entire genome
    cost = cost of having multisegments
    w = fitness
    progeny_n = number of progenies a virus agent will have during reproduction. Default is 0.
    """
    def __init__(self,k1, k2):
    	self.id = 2
        self.k1 = k1
        self.k2 = k2
        self.k = self.k1 + self.k2
        self.progeny_n = 0
    
    def mutate(self,mu):
        """
        Mutation in sequence before reproduction
        mu = mutation rate
        """
        mut_num = int(np.random.binomial(L, mu)) # number of mutation
        if not back:
            split_pt = int(np.ceil(np.random.uniform(0,1)*mut_num)) # how much of the mutation segment1 is getting
            self.k1 += split_pt
            self.k2 += mut_num - split_pt
        else:
            for i in range(mut_num):
                p = np.random.uniform(0,1)
                if np.random.uniform(0,L) < self.k: # back mutation
                    if self.k1 == 0:
                        self.k2 -= 1
                        self.k -= 1
                    elif self.k2 == 0:
                        self.k1 -= 1
                        self.k -= 1
                    else: 
                        if p < 0.5: # seg1
                            self.k1 -= 1
                            self.k -= 1
                        else: # seg2
                            self.k2 -= 1
                            self.k -= 1
                else: # normal mutation
                    if p < 0.5: # seg1
                        self.k1 += 1
                        self.k += 1
                    else: # seg2
                        self.k2 += 1
                        self.k += 1

def step(pop):
	while len(next_gen) < N0:
		next_gen = []
		popsize = [0,0]	
		sample = np.random.choice(pop, 2, replace=False)
		if sample[0].id == 1 or sample[1].id == 1:
			if np.random.uniform(0,1) < 0.5: # sample 0 or 1
				if np.random.uniform(0,1) < (1-s)**sample[0].k: # progeny live or die
					if sample[0].id == 1:
						next_gen.append(Virus1(sample[0].k))
						popsize[0] += 1
					else:
						next_gen.append(Virus2(sample[0].k1,sample[0].k2))
						popsize[1] += 1
			else:
				if np.random.uniform(0,1) > (1-s)**sample[1].k:
					if sample[1].id == 1:
						next_gen.append(Virus1(sample[1].k))
						popsize[0] += 1
					else:
						next_gen.append(Virus2(sample[1].k1,sample[1].k2))
						popsize[1] += 1
		elif sample[0].id == 2 and sample[1].id == 2:
			if np.random.uniform(0,1) < 0.5:
				if np.random.uniform(0,1) > (1-s)**(sample[0].k1+sample[1].k2):
					next_gen.append(Virus2(sample[0].k1,sample[1].k2))
					popsize[1] += 1
			else:
				if np.random.uniform(0,1) > (1-s)**(sample[0].k2+sample[1].k1):
					next_gen.append(Virus2(sample[0].k2,sample[1].k1))
					popsize[1] += 1

	if np.sum(popsize) != N0:
		raise ValueError('popsize doesn\'t add up to N0!!')
	return next_gen, popsize


start = timeit.default_timer() # timer start
bar = Bar('Processing', max=rep) # progress bar start
# write out data with file name indicating time it started collecting
now = datetime.datetime.now()
destination = 'test'
if len(sys.argv) > 1:
    try:
        destination = sys.argv[1] 
    except:
        pass
if destination not in os.listdir('./data'):
    os.system('mkdir ./data/' + destination)
params = '%d,%d,%d,%.2f,%d,%d,%.5f,%d,%.2f,%.2f,%.2f'%(back,rep,L,s,N0,K,mu,gen_num,cost,r,N1r)
tail = 'c1.3s_%s(0).csv'%(params)
while tail in os.listdir('./data/'+destination):
    lastnum = int(tail[-6])
    tail = tail[0:-6] + str(lastnum+1) + tail[-5::] 
file_name = './data/' + destination + '/' + tail
fh = open(file_name,'w')

if timestep:
    fh.write('rep,t,pop1,pop2,k1,k2\n')
    for repe in range(rep):
		# initiate
		for i in N0*N1r:
			pop.append(Virus1(0))
		for i in N0*(1-N1r):
			pop.append(Virus2(0,0))
		popsize = [N0*N1r,N0*(1-N1r)]

		# run through generation
		for gen in gen_num:
			for i in range(len(pop)):
				pop[i].mutate(mu)
			pop, popsize = step(pop)
            # recording k
            ks1 = [] # k's for each virus in a subpop
            ks2 = []
            for i in range(len(pop)):
            	if pop[i].id == 1:
            		ks1.append(pop[i].k)
            	else:
            		ks2.append(pop[i].k)
            if krecord == 1:                	
                if len(ks1) == 0:
                	ks1.append('NA')
                if len(ks2) == 0:
                	ks2.append('NA')
                ks1 = str(ks1).replace(', ','.')[1:-1]
                ks2 = str(ks2).replace(', ','.')[1:-1]
                fh.write('%d,%d,%d,%d,%s,%s\n'%(repe+1,gen+1,popsize[0],popsize[1],ks1,ks2))
            elif krecord == 0:
                if len(ks1) == 0:
                	ks1 = -1
                else:
                	ks1 = np.mean(np.array(ks1))
                if len(ks2) == 0:
                	ks2 = -1
                else krecord == 0:
                	ks2 = np.mean(np.array(ks2))
                fh.write('%d,%d,%d,%d,%.2f,%.2f\n'%(repe+1,gen+1,popsize[0],popsize[1],ks1,ks2))
            elif krecord == 2:
                if len(ks1) == 0:
                	ks1 = -1
                else:
                	ks1 = np.min(np.array(ks1))
                if len(ks2) == 0:
                	ks2 = -1
                else krecord == 0:
                	ks2 = np.min(np.array(ks2))
                fh.write('%d,%d,%d,%d,%d,%d\n'%(repe+1,gen+1,popsize[0],popsize[1],ks1, ks2))
        bar.next()

else:
    fh.write('pop1,pop2,k1,k2\n')
    for repe in range(rep):
		# initiate
		for i in N0*N1r:
			pop.append(Virus1(0))
		for i in N0*(1-N1r):
			pop.append(Virus2(0,0))
		popsize = [N0*N1r,N0*(1-N1r)]

		# run through generation
		for gen in gen_num:
			for i in range(len(pop)):
				pop[i].mutate(mu)
			pop, popsize = step(pop)

        # recording k
        ks1 = [] # k's for each virus in a subpop
        ks2 = []
        for i in range(len(pop)):
        	if pop[i].id == 1:
        		ks1.append(pop[i].k)
        	else:
        		ks2.append(pop[i].k)
        if krecord == 1:                	
            if len(ks1) == 0:
            	ks1.append('NA')
            if len(ks2) == 0:
            	ks2.append('NA')
            ks1 = str(ks1).replace(', ','.')[1:-1]
            ks2 = str(ks2).replace(', ','.')[1:-1]
            fh.write('%d,%d,%s,%s\n'%(popsize[0],popsize[1],ks1,ks2))
        elif krecord == 0:
            if len(ks1) == 0:
            	ks1 = -1
            else:
            	ks1 = np.mean(np.array(ks1))
            if len(ks2) == 0:
            	ks2 = -1
            else krecord == 0:
            	ks2 = np.mean(np.array(ks2))
            fh.write('%d,%d,%.2f,%.2f\n'%(popsize[0],popsize[1],ks1,ks2))
        elif krecord == 2:
            if len(ks1) == 0:
            	ks1 = -1
            else:
            	ks1 = np.min(np.array(ks1))
            if len(ks2) == 0:
            	ks2 = -1
            else krecord == 0:
            	ks2 = np.min(np.array(ks2))
            fh.write('%d,%d,%d,%d\n'%(popsize[0],popsize[1],ks1, ks2))
        bar.next()

fh.close()
stop = timeit.default_timer()
print('\nthe simulation ran for %.2f min'%((stop - start)/60))
bar.finish()

File: Reassortment/comp1.3.c
/*
 Influenza Competition model 1.3
 competition model btw 1seg and 2seg in a WF model
 Some key differences from 1.2's:
 - WF model (no demographic factor or carrying capcacity for that matter.)
 Reproduction process:
 1. pick 2 random parents
 2. If either one is 1segment, only one of them gets replicated.
 3. If both of them are 2segement, it goes through recombination with 
    a probability of r, and one of the progeny is created.
 4. the created progeny from either proces  2. or 3. are assessed 
 5. whether they will survive by their fitness factor w.
 6. If a random number ([0,1]) exceeds w, the progeny dies and 
    has to go through the process 2.-5. again.
*/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

/*
 Parameters
 back = whether the simulation allows back mutation (1 if there's one. 0 if there isn't)
 timestep = whether a simulation wants to record every generation or only the end generation per repetition
 krecord = how you want to record k in the output csv. 0= mean k. 1=all k values of indivs in an array. 2= minimum k.
 N0 = Population size
 K = Carrying capacity
 L = sequence length
 s = fitness decrease from deleterious mutation
 mu = mutation rate per site
 gen_num = generation amount 
 r = reassortment rate
 rep = repetition amount
 N1r = ratio of 1segment virus
*/

int back = 0;
int timestep = 0;
int krecord = 0;
char *destination = "ctest";
int rep = 1;
int L = 300;
double s = 0.05;
#define N0 50
int K = 1000;
float mu = 0.005;
int gen_num = 10;
double cost = 0.00;
double r = 0.5;
double N1r = 0.5;
long seed = -1;

struct virus {
	int id;
	int k1;
	int k2;
	int k;
};

//https://stackoverflow.com/questions/2620146/how-do-i-return-multiple-values-from-a-function-in-c
float ran1(long *seed);
float gammln(float xx);
float bnldev(float pp, int n, long *idum);
struct virus *step(struct virus popop[],struct virus *next_gen_p);


int main(void) {
	//initialize pop
	struct virus pop[N0];
	int N1 = N0*N1r;
	int N2 = N0*(1-N1r);
	for (int i=0;i<N1;i++){
		pop[i].id = 1;
		pop[i].k1 = 0;
		pop[i].k2 = 0;
		pop[i].k = 0;
	}
	for (int i=0;i<N2;i++){
		pop[i+N1].id = 2;
		pop[i+N1].k1 = 0;
		pop[i+N1].k2 = 0;
		pop[i+N1].k = 0;	
	}

	// set progress bar, initiate csv file, and start timer


	// simulation start
	struct virus next_gen[N0];
	struct virus *pop2;
	for (int repe=0; repe<rep; repe++){	
		for (int i=0; i<10; i++){ // run through generation
			mutate(pop);
			pop2 = step(pop,&next_gen);
			memcpy(pop,pop2,sizeof(struct virus)*N0); // cycle between pop and pop2 to continue looping.
			printf("pop0 outside of step id:%d, k1:%d, k2:%d, k:%d\n",pop[0].id,pop[0].k1,pop[0].k2,pop[0].k);
			printf("pop1 outside of step id:%d, k1:%d, k2:%d, k:%d\n",pop[1].id,pop[1].k1,pop[1].k2,pop[1].k);
			printf("pop2 outside of step id:%d, k1:%d, k2:%d, k:%d\n",pop[2].id,pop[2].k1,pop[2].k2,pop[2].k);
			printf("\n\n");
			//record
		}
	return 0;
}












void record(struct virus popop[]) {
	if (timestep) {

	} else {

	} 
}




void mutate(struct virus popop[]) {
	// goes through population and make every inidividual go through 
	// mutation process.
	// input: population struct array.
	// output: population struct with updated k values according to
	//   mutation rate.
	for (int i=0;i<N0; i++){
		int mut_num = bnldev(mu,L,&seed); // binomial pick of number of mutation based on mu.
		if (back == 1){ // back mutation
			if (popop[i].id == 1) { // individual is 1segment
				for (int i=0; i<mut_num; i++) { // add or subtract k
					if (ran1(&seed) < popop[i].k/L) { // k/L is the probability of back mutating.
						popop[i].k -= 1;
					} else {
						popop[i].k += 1;
					}
				}
			} else { // individual is 2segment
				for (int i=0; i<mut_num; i++) { // add or subtract k1 or k2, and k.
					if (ran1(&seed) < popop[i].k/L) {
						if (popop[i].k1==0) {
							popop[i].k2 -= 1;
							popop[i].k -= 1;
						} else if (popop[i].k2 == 0) {
							popop[i].k1 -= 1;
							popop[i].k -= 1;
						} else {
							if (ran1(&seed) < 0.5) {
								popop[i].k1 -= 1;
								popop[i].k -= 1;
							} else {
								popop[i].k2 -= 1;
								popop[i].k -= 1;
							}
						}
					} else {
						if (ran1(&seed) < 0.5) {
							popop[i].k1 += 1;
							popop[i].k += 1;
						} else {
							popop[i].k2 += 1;
							popop[i].k += 1;
						}
					}
				}
			}
		} else { // no back mutation 
			if (popop[i].id == 1) {
				popop[i].k += mut_num;
			} else {
				int breakpt = floor(ran1(&seed)*(mut_num+1));
				popop[i].k1 += breakpt;
				popop[i].k2 += mut_num - breakpt;
				popop[i].k = popop[i].k1 + popop[i].k2;
			}
		}
	}
}

struct virus *step(struct virus popop[],struct virus *next_gen_p) {
	// goes through reproduction process
	// the process is depicted at the top of the script.
	// input: pop struct array
	// output: next generation's pop struct array
	int l = 0; // next gen length
	while (l < N0) {
		int s1 = floor(ran1(&seed)*N0); // sample 1
		int s2 = floor(ran1(&seed)*N0); // sample 2
		//printf("s1:%d k1:%d k2:%d k:%d\n",s1,popop[s1].k1,popop[s1].k2,popop[s1].k);
		//printf("s2:%d k1:%d k2:%d k:%d\n",s2,popop[s2].k1,popop[s2].k2,popop[s2].k);
		if (popop[s1].id == 1 || popop[s2].id == 1) { // either parents is segment 1
			if (ran1(&seed) < 0.5) { // pick s1
				if (ran1(&seed) < pow(1.0-s,popop[s1].k)){
					next_gen_p[l].id = popop[s1].id;
					next_gen_p[l].k1 = popop[s1].k1;
					next_gen_p[l].k2 = popop[s1].k2;
					next_gen_p[l].k = popop[s1].k;
					l += 1;
				}
			} else { // pick s2
				if (ran1(&seed) < pow(1.0-s,popop[s2].k)){
					next_gen_p[l].id = popop[s2].id;
					next_gen_p[l].k1 = popop[s2].k1;
					next_gen_p[l].k2 = popop[s2].k2;
					next_gen_p[l].k = popop[s2].k;
					l += 1;
				}
			}
		} else { // both parents segment 2
			if (ran1(&seed) < 0.5) { // pick k1 from s1 and k2 from s2
				if (ran1(&seed) < pow(1.0-s,(popop[s1].k1 + popop[s2].k2))){
					next_gen_p[l].id = popop[s1].id;
					next_gen_p[l].k1 = popop[s1].k1;
					next_gen_p[l].k2 = popop[s2].k2;
					next_gen_p[l].k = popop[s1].k;
					l += 1;
				}
			} else { // pick k1 from s2 and k2 from s1
				if (ran1(&seed) < pow(1.0-s,(popop[s1].k2 + popop[s2].k1))){
					next_gen_p[l].id = popop[s2].id;
					next_gen_p[l].k1 = popop[s2].k1;
					next_gen_p[l].k2 = popop[s1].k2;
					next_gen_p[l].k = popop[s2].k;
					l += 1;
				}
			}
		}
	}
	printf("pop0 inside of step id:%d, k1:%d, k2:%d, k:%d\n",next_gen_p[0].id,next_gen_p[0].k1,next_gen_p[0].k2,next_gen_p[0].k);
	printf("pop1 inside of step id:%d, k1:%d, k2:%d, k:%d\n",next_gen_p[1].id,next_gen_p[1].k1,next_gen_p[1].k2,next_gen_p[1].k);
	printf("pop2 inside of step id:%d, k1:%d, k2:%d, k:%d\n",next_gen_p[2].id,next_gen_p[2].k1,next_gen_p[2].k2,next_gen_p[2].k);
	printf("\n\n");
	return next_gen_p;
}





// random number generating functions ran1=uniform [0,1], bnldev= binomial. gammln is needed for bnldev.
#define IA 16807
#define IM 2147483647
#define AM (1.0/IM)
#define IQ 127773
#define IR 2836
#define NTAB 32
#define NDIV (1+(IM-1)/NTAB)
#define EPS 1.2e-7
#define RNMX (1.0-EPS)
float ran1(long *idum)
{
  int j;
  long k;
  static long iy=0;
  static long iv[NTAB];
  float temp;

  if (*idum <= 0 || !iy) {
    if (-(*idum) < 1) *idum=1;
    else *idum = -(*idum);
    for (j=NTAB+7;j>=0;j--) {
      k=(*idum)/IQ;
      *idum=IA*(*idum-k*IQ)-IR*k;
      if (*idum < 0) *idum += IM;
      if (j < NTAB) iv[j] = *idum;
    }
    iy=iv[0];
  }
  k=(*idum)/IQ;
  *idum=IA*(*idum-k*IQ)-IR*k;
  if (*idum < 0) *idum += IM;
  j=iy/NDIV;
  iy=iv[j];
  iv[j] = *idum;
  if ((temp=AM*iy) > RNMX) return RNMX;
  else return temp;
}
#undef IA
#undef IM
#undef AM
#undef IQ
#undef IR
#undef NTAB
#undef NDIV
#undef EPS
#undef RNMX

float gammln(float xx)
{
	double x,y,tmp,ser;
	static double cof[6]={76.18009172947146,-86.50532032941677,
		24.01409824083091,-1.231739572450155,
		0.1208650973866179e-2,-0.5395239384953e-5};
	int j;

	y=x=xx;
	tmp=x+5.5;
	tmp -= (x+0.5)*log(tmp);
	ser=1.000000000190015;
	for (j=0;j<=5;j++) ser += cof[j]/++y;
	return -tmp+log(2.5066282746310005*ser/x);
}

#define PI 3.141592654
float bnldev(float pp, int n,long *idum)
{
	float gammln(float xx);
	float ran1(long *idum);
	int j;
	static int nold=(-1);
	float am,em,g,angle,p,bnl,sq,t,y;
	static float pold=(-1.0),pc,plog,pclog,en,oldg;

	p=(pp <= 0.5 ? pp : 1.0-pp);
	am=n*p;
	if (n < 25) {
		bnl=0.0;
		for (j=1;j<=n;j++)
			if (ran1(idum) < p) bnl += 1.0;
	} else if (am < 1.0) {
		g=exp(-am);
		t=1.0;
		for (j=0;j<=n;j++) {
			t *= ran1(idum);
			if (t < g) break;
		}
		bnl=(j <= n ? j : n);
	} else {
		if (n != nold) {
			en=n;
			oldg=gammln(en+1.0);
			nold=n;
		} if (p != pold) {
			pc=1.0-p;
			plog=log(p);
			pclog=log(pc);
			pold=p;
		}
		sq=sqrt(2.0*am*pc);
		do {
			do {
				angle=PI*ran1(idum);
				y=tan(angle);
				em=sq*y+am;
			} while (em < 0.0 || em >= (en+1.0));
			em=floor(em);
			t=1.2*sq*(1.0+y*y)*exp(oldg-gammln(em+1.0)
				-gammln(en-em+1.0)+em*plog+(en-em)*pclog);
		} while (ran1(idum) > t);
		bnl=em;
	}
	if (p != pp) bnl=n-bnl;
	return bnl;
}
#undef PI
